# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webscheduling/WebTaskSchedulerWorker.h
# Commit: 4739af17e757
# Full Hash: 4739af17e757724c0833a8564579833249389566
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2025-02-27 21:52:52
# Regressor Bug: 1920115
# File Overlap Count: 2
# Description:
#   Bug 1920115 - Allow high priority tasks to run before timers r=smaug
#   
#   In Gecko, unlike other runnables where the TaskController
#   picks one each at a time, the TimeoutManager allows multiple
#   timers to run sequentially without allowing other runnables
# ==============================================================================

diff -r 0773f669c498 -r 4739af17e757 dom/webscheduling/WebTaskSchedulerWorker.h
--- a/dom/webscheduling/WebTaskSchedulerWorker.h	Thu Feb 27 16:17:36 2025 +0000
+++ b/dom/webscheduling/WebTaskSchedulerWorker.h	Thu Feb 27 16:17:36 2025 +0000
@@ -38,6 +38,13 @@
 
   void Disconnect() override;
 
+  void IncreaseNumNormalOrHighPriorityQueuesHaveTaskScheduled() override;
+  void DecreaseNumNormalOrHighPriorityQueuesHaveTaskScheduled() override;
+
+  bool HasScheduledNormalOrHighPriorityWebTasks() const {
+    return mNumHighPriorityQueuesHaveTaskScheduled;
+  }
+
  private:
   ~WebTaskSchedulerWorker() = default;
 
@@ -47,6 +54,12 @@
 
   RefPtr<StrongWorkerRef> mWorkerRef;
   bool mWorkerIsShuttingDown{false};
+
+  // Unlike window global where multiple globals can share the
+  // same event loop, worker globals don't share event loops,
+  // so it's okay to have this counter lives inside the
+  // scheduler for workers.
+  uint32_t mNumHighPriorityQueuesHaveTaskScheduled = 0;
 };
 }  // namespace mozilla::dom
 #endif